### PR TITLE
fix: add v8 snapshot usage to cypress in cypress testing

### PR DESCRIPTION
**Pull request by ryanthemanuel**
### User facing changelog

n/a

### Additional details

This adds support for utilizing the v8 snapshot in cypress in cypress testing. In order to do this, we need to ensure that when the child process is started it is done via the `spawn` command since electron has [special logic](https://github.com/electron/electron/blob/main/patches/node/refactor_alter_child_process_fork_to_use_execute_script_with.patch) in its `fork` command which effectively disables the snapshot. This has the implication of having an additional Cypress icon in the dock on Macs, but that will only happen during development and is a small inconvenience given that we will be able to more accurately simulate prod during cypress in cypress tests.

In addition, as just general cleanup, I cleaned up the `bytenode` logic and was able to make it a dev dependency by bundling it into the electron entry point

### Steps to test

Running the app and launchpad tests is the best way to see this in action

### How has the user experience changed?

n/a

### PR Tasks

- [X] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/data-context/src/data/ProjectConfigIpc.ts [hunk 1]
```diff
@@ -310,6 +319,17 @@ export class ProjectConfigIpc extends EventEmitter {
...
+      configProcessArgs.push('--no-sandbox')
+    }
+
+    if (process.env.CYPRESS_INTERNAL_E2E_TESTING_SELF_PARENT_PROJECT) {
+      return spawn(process.execPath, ['--entryPoint', CHILD_PROCESS_FILE_PATH, ...configProcessArgs], {
+        ...childOptions,
+        stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
+      })
+    }
+
```
Comment (ryanthemanuel) [inline packages/data-context/src/data/ProjectConfigIpc.ts]:
```suggestion
    if (process.env.CYPRESS_INTERNAL_E2E_TESTING_SELF_PARENT_PROJECT) {
      if (isSandboxNeeded()) {
        configProcessArgs.push('--no-sandbox')
      }

      return spawn(process.execPath, ['--entryPoint', CHILD_PROCESS_FILE_PATH, ...configProcessArgs], {
        ...childOptions,
        stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
      })
    }

```

File: packages/config/package.json [hunk 2]
```diff
@@ -21,7 +21,7 @@
     "@babel/parser": "^7",
     "@babel/plugin-syntax-typescript": "^7",
     "@babel/plugin-transform-typescript": "^7",
-    "@babel/traverse": "^7",
+    "@babel/traverse": "7.15.4",
```
Comment (ryanthemanuel) [inline packages/config/package.json:24]: Updated this to match the rest of the repo.

File: packages/server/index.js [hunk 3]
```diff
@@ -10,10 +15,16 @@ const startCypress = async () => {
 
     await require('./start-cypress')
   } catch (error) {
-    // eslint-disable-next-line no-console
+  // eslint-disable-next-line no-console
```
Comment (ryanthemanuel) [inline packages/server/index.js]:
```suggestion
    // eslint-disable-next-line no-console
```

**Review by AtofStryker — APPROVED** [review]

File: scripts/after-pack-hook.js [hunk 4]
```diff
@@ -10,67 +10,74 @@ const { buildEntryPointAndCleanup } = require('./binary/binary-cleanup')
...
-  })
-
-  const buildSubfoldersPerPlatform = {
-    darwin: join('Cypress.app', 'Contents', 'Resources', 'app'),
-    linux: join('resources', 'app'),
-    win32: join('resources', 'app'), // TODO check this path
-  }
-  const buildSubfolder = buildSubfoldersPerPlatform[os.platform()]
-  const outputFolder = join(params.appOutDir, buildSubfolder)
+  try {
```
Comment (flotwig) [inline scripts/after-pack-hook.js:13]: I wouldn't have expected test changes to affect build scripts, why are these changes needed?

Comment (ryanthemanuel) [reply scripts/after-pack-hook.js:13]: Oops forgot to comment on the description for why this was changed. I added a couple of changes to be able to move `bytenode` from a main dependency to a dev dependency and just generally cleaned some of this up. No functional changes, but it was something that was nagging at me from my last PR that I wanted to clean up while I had time.

**Review by flotwig — APPROVED** [review]

File: scripts/after-pack-hook.js (hunk 4 above)
Comment (ryanthemanuel) [reply scripts/after-pack-hook.js:13]: Added this to the description.

File: packages/data-context/src/data/ProjectConfigIpc.ts [hunk 5]
```diff
@@ -310,6 +319,17 @@ export class ProjectConfigIpc extends EventEmitter {
       debug(`no typescript found, just use regular Node.js`)
     }
 
+    if (process.env.CYPRESS_INTERNAL_E2E_TESTING_SELF_PARENT_PROJECT) {
```
Comment (lmiller1990) [inline packages/data-context/src/data/ProjectConfigIpc.ts:322]: Bit of a shame we need to ship test code to prod :'(

This is "the change" that lets v8 snapshots run in cy-in-cy?

Comment (lmiller1990) [reply packages/data-context/src/data/ProjectConfigIpc.ts:322]: Is there anyway to verify it now uses snapshots in the tests? How do we know it works (and catch it if stops?)

**Review by lmiller1990 — APPROVED** [review]
Great, I have a question, but no blocker.

File: packages/data-context/src/data/ProjectConfigIpc.ts (hunk 5 above)
Comment (ryanthemanuel) [reply packages/data-context/src/data/ProjectConfigIpc.ts:322]: Yeah it's unfortunate, but I'm not sure of a clean way to do it otherwise with how everything is structured.

And, I like your idea. I added a check to ensure that we fail in cypress in cypress if the snapshot is unavailable.
